CRM/Core/BAO/Translation.php

```diff
@@ -218,12 +218,12 @@ protected static function getTranslatedFieldsForRequest(AbstractAction $apiReque
       ->addWhere('entity_table', '=', CRM_Core_DAO_AllCoreTables::getTableForEntityName($apiRequest['entity']))
       ->setCheckPermissions(FALSE)
       ->setSelect(['entity_field', 'entity_id', 'string', 'language']);
-    if ((substr($userLocale->nominal, '-3', '3') !== '_NO')) {
+    if ((substr($userLocale->nominal ?? '', '-3', '3') !== '_NO')) {
       // Generally we want to check for any translations of the base language
       // and prefer, for example, French French over US English for French Canadians.
       // Sites that genuinely want to cater to both will add translations for both
       // and we work through preferences below.
-      $translations->addWhere('language', 'LIKE', substr($userLocale->nominal, 0, 2) . '%');
+      $translations->addWhere('language', 'LIKE', substr($userLocale->nominal ?? '', 0, 2) . '%');
```

Comment:
The first one above is ok in the context of "is it norwegian or not", but for this one wouldn't it then return EVERY language. Is this fail actually hiliting something we should be looking at, i.e. was the test passing before just because there is only en_US? Which test?

Reply:
Oh I see you have listed which tests.

Reply:
the tests that were failing are all in CRM_Core_BAO_MessageTemplateTest

Reply:
also on my php 7.4 I get this if we assume $userlocale->nominal is NULL so I think this is equivalent 

```
seamus@civicrmdevelopment:~/buildkit/build/47-test/web/sites/all/modules/civicrm$ drush ev "var_dump(substr(NULL, 0, 2))"
string(0) ""
```

Reply:
It's equivalent but I'm looking at whether it was correct before. I'm not sure what it means when tsLocale (aka nominal) is null.

Reply:
I'm conflicted. It's difficult to reproduce on a real site because I think on a real site it _is_ an error that there is no language set. So the test is probably not mimicking a real site, in which case this error seems more like it's flagging something that needs attention. But having said that, scripts or other api usage may not mimick a real site either and that's not necessarily wrong.

The Locale object seems to have been coded with null being a possibility, but this function here seems to have been coded as if it's never going to be null, so that's the conflict. It ends up returning all languages and then picks one. That seems semi-wrong, and if it's just silenced it will probably stay semi-wrong. Priorities I guess. I think I'll merge and then just add a TODO code comment.